File: lc/c/023.c
bool precheck(char **board, int numRows, int numCols, char *word) {
    int cnt[128] = {0};
    for (int i = 0; i < strlen(word); ++i) {
        cnt[word[i]] += 1;
    }
    for (int i = 0; i < numRows; ++i){
        for (int j = 0; j < numCols; ++j) {
            if (cnt[board[i][j]])
                cnt[board[i][j]] -= 1;
        }
    }
    for (int i = 0; i < sizeof(cnt)/sizeof(*cnt); ++i) {
        if (cnt[i] > 0)
            return false;
    }
    return true;
}

bool dfs(char **board, int numRows, int numCols,
         int r, int c, char *pos) {
    if (*pos == 0)
        return true;
    else if (r < 0 || c < 0 || r >= numRows || c >= numCols)
        return false;
    else if (board[r][c] != *pos)
        return false;

    board[r][c] |= 0x80; // visited
    int dir[4][2] = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    for (int i = 0; i < sizeof(dir)/sizeof(*dir); ++i) {
        int x = dir[i][0], y = dir[i][1];
        if (dfs(board, numRows, numCols, r+x, c+y, pos+1))
            return true;
    }
    board[r][c] &= ~0x80; // unvisited
    return false;
}

bool exist(char** board, int boardSize, int* boardColSize, char * word){
    if (!precheck(board, boardSize, boardColSize[0], word))
        return false;

    for (int r = 0; r < boardSize; ++r) {
        for (int c = 0; c < boardColSize[r]; ++c) {
            if (dfs(board, boardSize, boardColSize[r], r, c, word))
                return true;
        }
    }
    return false;
}

File: lc/py/023.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def precheck():
            counter = collections.Counter(word)
            for row in board:
                for ch in row:
                    if ch in counter:
                        counter[ch] -= 1

            for count in counter.values():
                if count > 0:
                    return False
            return True

        def dfs(r, c, pos):
            if pos == len(word):
                return True
            elif r < 0 or c < 0 or r >= len(board) or c >= len(board[r]):
                return False
            elif (r, c) in visited:
                return False
            elif board[r][c] != word[pos]:
                return False

            visited.add((r, c))
            for x, y in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                if dfs(r+x, c+y, pos+1):
                    return True
            visited.remove((r,c))
            return False

        if not precheck():
            return False

        visited = set()
        for r in range(len(board)):
            for c in range(len(board[r])):
                if dfs(r, c, 0):
                    return True
        return False
